Add Expressive Code to Starlight (#742)
withastro/starlight: docs/src/content/docs/guides/authoring-content.md

```
````

### Expressive Code features

Starlight uses [Expressive Code](https://github.com/expressive-code/expressive-code/tree/main/packages/astro-expressive-code) to extend formatting possibilities for code blocks.
Expressive Code’s text markers and window frames plugins are enabled by default.
Code block rendering can be configured using Starlight’s [`expressiveCode` configuration option](/reference/configuration/#expressivecode).

> #### Text markers

Reviewer: How standard is the phrase "text markers" for what seems to be describing "highlighting"? I don't think I use this phrase myself at all, so I might be tempted to rename?

```suggestion
#### Line and Text highlighting
```

Reviewer 2: I also considered this — and “highlight” is used as a verb in this section. I think the slight difficulty is that “syntax highlighting” has such a specific meaning that I wasn’t sure if “highlighting” in this context might imply _that_, rather than what we might think of as “highlighting” in a word processing or handwritten context. Do you think that could be an issue?

Author: I went through the same struggle when naming this feature and the plugin providing it. The unwanted proximity to the widely known and used term "syntax highlighting" was the deciding factor for me that I had to find another name, so I ended up with "text markers".

> You can highlight specific lines or parts of your code blocks using text markers.

Reviewer: ```suggestion
You can highlight specific lines or parts of your code blocks using [Expressive Code text markers](https://github.com/expressive-code/expressive-code/blob/main/packages/%40expressive-code/plugin-text-markers/README.md#usage-in-markdown--mdx-documents) on the opening line of your code block.
Use curly braces (`{ }`)  to highlight entire lines, and quotation marks to highlight strings of text.
```

> There are three styles of marker: a default neutral colored highlight, a green highlight indicating code that was inserted, and a red highlight indicating code that was deleted.

Reviewer: ```suggestion

There are three highlighting styles: neutral for calling attention to code, green for indicating inserted code, and red for indicating deleted code.
Both text and entire lines can be marked using the default marker, or in combination with `ins=` and `del=` to produce the desired highlighting.
```
I've put each sentence on its own line in this section, but I doubt this should all be one paragraph, so there's a line break at the top. Obvs, do what makes visual sense here.

> - [Mark entire lines & line ranges](https://github.com/expressive-code/expressive-code/blob/main/packages/%40expressive-code/plugin-text-markers/README.md#marking-entire-lines--line-ranges):

Reviewer: ```suggestion
Expressive Code provides several options for customizing the visual appearance of your code samples. 
Many of these can be combined, for highly illustrative code samples.
Please explore the [Expressive Code documentation](https://github.com/expressive-code/expressive-code/blob/main/packages/%40expressive-code/plugin-text-markers/README.md) for the extensive options available.
Some of the most common examples are shown below: 

- [Mark entire lines & line ranges using the `{ }` marker](https://github.com/expressive-code/expressive-code/blob/main/packages/%40expressive-code/plugin-text-markers/README.md#marking-entire-lines--line-ranges):
```

>   ```js {2-3}

Reviewer: I have a medium-strong preference for swapping all these examples, and showing the code you need to write first, and *then* what it looks like.

Feels more in keeping with helping people write something: see the code, THEN see the results it produces.

Reviewer 2: I also do! The only issue is that the entire page currently uses this ordering so I felt unsure about changing this without also revisiting the whole page, which felt like something that should maybe be done in a dedicated PR?

Author: I agree with both of you! I also see value in swapping the examples, but we should do it in a separate PR after getting this one over the finish line. :)

  function demo() {
    // This line (#2) and the next one are highlighted
    return 'This is line #3 of this snippet';
  }
  ```

  ````md
  ```js {2-3}
  function demo() {
    // This line (#2) and the next one are highlighted
    return 'This is line #3 of this snippet';
  }
  ```
  ````

> - [Mark selections of text](https://github.com/expressive-code/expressive-code/blob/main/packages/%40expressive-code/plugin-text-markers/README.md#marking-entire-lines--line-ranges):

Reviewer: ```suggestion
- [Mark selections of text using the `" "` marker or regular expressions](https://github.com/expressive-code/expressive-code/blob/main/packages/%40expressive-code/plugin-text-markers/README.md#marking-entire-lines--line-ranges):
```

  ```js "Individual terms" /Even.*supported/
  // Individual terms can be highlighted, too
  function demo() {
    return 'Even regular expressions are supported';
  }
  ```

  ````md
  ```js "Individual terms" /Even.*supported/
  // Individual terms can be highlighted, too
  function demo() {
    return 'Even regular expressions are supported';
  }
  ```
  ````

> - [Mark text as inserted or deleted with `ins` or `del`](https://github.com/expressive-code/expressive-code/blob/main/packages/%40expressive-code/plugin-text-markers/README.md#selecting-marker-types-mark-ins-del):

Reviewer: ```suggestion
- [Mark text or lines as inserted or deleted with `ins` or `del`](https://github.com/expressive-code/expressive-code/blob/main/packages/%40expressive-code/plugin-text-markers/README.md#selecting-marker-types-mark-ins-del):
```

  ```js "return true;" ins="inserted" del="deleted"
  function demo() {
    console.log('These are inserted and deleted marker types');
    // The return statement uses the default marker type
    return true;
  }
  ```

  ````md
  ```js "return true;" ins="inserted" del="deleted"
  function demo() {
    console.log('These are inserted and deleted marker types');
    // The return statement uses the default marker type
    return true;
  }
  ```
  ````

> - [Combine syntax highlighting with `diff`-like syntax](https://github.com/expressive-code/expressive-code/blob/main/packages/%40expressive-code/plugin-text-markers/README.md#combining-syntax-highlighting-with-diff-like-syntax):

Reviewer: ```suggestion
- [Combine language-based syntax highlighting with `diff`-like syntax](https://github.com/expressive-code/expressive-code/blob/main/packages/%40expressive-code/plugin-text-markers/README.md#combining-syntax-highlighting-with-diff-like-syntax):
```

OOh, question, does this mean that for Starlight, we do NOT do ins/del for individual lines? I know this is the solution we use for standard Markdown, but our normal `ins={2-4}` should work in Starlight, right? Does this mean you are suggesting this notation (for universal compatibility?)

If so, I might further explain this not as the first resort, but as a way to write code samples in such a way that they'll render in non-Expressive Code environments as diffs, but in your Starlight site, they can take advantage additionally of language highlighting.  (In part, because our code samples in Docs are typically going to use the `ins={2-4}` notation, and so the docs here should maybe have an example of that?)

Reviewer 2: I believe both diff-style and `ins={...}` should work! So it’s a matter of taste which you choose (although I think the diff-style is probably preferable in most cases for a couple of reasons:

1. Diff highlighting works everywhere (same reason we were using it in Astro integration READMEs).
2. You don’t need to count and maintain line numbers. If you use `+`/`-` at the start of lines, adding or moving parts of the code doesn’t require you to manually update the `ins`/`del` ranges.

Author: They do both still work, yes. Nothing has been removed. I just tried to limit the amount of EC-focused examples included in the Starlight docs to a reasonable amount and therefore decided to just show an example of the diff-like syntax for the reasons Chris mentioned.

Reviewer 2: Yes, I think once Expressive Code has its own docs to show these off, we may even be able to reduce the examples in favour of linking there.

  function demo() {
    console.log('These are inserted and deleted marker types');
    // The return statement uses the default marker type
    return true;
  };
  ```

  ````md
  ```js "return true;" ins="inserted" del="deleted"
  function demo() {
    console.log('These are inserted and deleted marker types');
    // The return statement uses the default marker type
    return true;
  };
  ```
  ````

- [Combining syntax highlighting with `diff`-like syntax](https://github.com/expressive-code/expressive-code/blob/main/packages/%40expressive-code/plugin-text-markers/README.md#combining-syntax-highlighting-with-diff-like-syntax)

  ```diff lang="js"
    function thisIsJavaScript() {
      // This entire block gets highlighted as JavaScript,
      // and we can still add diff markers to it!
  -   console.log('Old code to be removed')
  +   console.log('New and shiny code!')
    }
  ```

  ````md
  ```diff lang="js"
    function thisIsJavaScript() {
      // This entire block gets highlighted as JavaScript,
      // and we can still add diff markers to it!
  -   console.log('Old code to be removed')
  +   console.log('New and shiny code!')
    }
  ```
  ````

### Window frames

> By default, every code block that has a language identifier is rendered inside a window frame. This is done by the frames plugin of Expressive Code, which is enabled by default in Starlight. Depending on the code's language identifier, the frame can look like a code editor (similar to VS Code), or like a terminal window.

Reviewer 2: This first sentence here is not strictly speaking true from what I see. Every shell language (`sh`, `bash` etc.) will be rendered in a Terminal-like frame, but other languages only get the tab window if they have a title, I think? Is that expected? I can update this in my docs PR if so.

Reviewer 2: And I guess the follow-up question would be when the `frame="none"` attribute becomes necessary? I guess mainly if a code comment would otherwise be used as the title?

Author: Oh, you're right! The code behaves as expected (and covered by many tests), I just wrote an incorrect statement in the accompanying documentation. Great that you found that! If you have the time to update this in the docs PR, that would be much appreciated!

Reviewer 2: Yes, done in https://github.com/hippotastic/starlight/pull/2!
